flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceSplitReader.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.io.CloseableIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class IcebergSourceSplitReader<T> implements SplitReader<RecordAndPosition<T>, IcebergSourceSplit> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSourceSplitReader.class);
+
+  private final ReaderFunction<T> readerFunction;
+  private final int indexOfSubtask;
+  private final IcebergSourceReaderMetrics metrics;
+  private final Queue<IcebergSourceSplit> splits;
+
+  private CloseableIterator<RecordsWithSplitIds<RecordAndPosition<T>>> currentReader;
+  private IcebergSourceSplit currentSplit;
+  private String currentSplitId;
+
+  IcebergSourceSplitReader(ReaderFunction<T> readerFunction,
+                           SourceReaderContext context,
+                           IcebergSourceReaderMetrics metrics) {
+    this.readerFunction = readerFunction;
+    this.indexOfSubtask = context.getIndexOfSubtask();
+    this.metrics = metrics;
+    this.splits = new ArrayDeque<>();
+  }
+
+  @Override
+  public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
+    metrics.incrementSplitReaderFetchCalls();
+    checkSplitOrStartNext();
+
+    if (currentReader.hasNext()) {
+      // Because Iterator#next() doesn't support checked exception,
+      // we need to wrap and unwrap the checked IOException with UncheckedIOException
+      try {
+        return currentReader.next();
+      } catch (UncheckedIOException e) {
+        throw e.getCause();
+      }
+    } else {
+      return finishSplit();
+    }
+  }
+
+  @Override
+  public void handleSplitsChanges(SplitsChange<IcebergSourceSplit> splitsChanges) {
+    LOG.info("Add splits to reader: {}", splitsChanges.splits());
+    splits.addAll(splitsChanges.splits());
+    metrics.incrementAssignedSplits(splitsChanges.splits().size());
+    metrics.incrementAssignedBytes(calculateBytes(splitsChanges));
+  }
+
+  @Override
+  public void wakeUp() {
+  }
+
+  @Override
+  public void close() throws Exception {
+    currentSplitId = null;
+    if (currentReader != null) {
+      currentReader.close();
+    }
+  }
+
+  private long calculateBytes(IcebergSourceSplit split) {
+    return split.task().files().stream()
+        .map(fileScanTask -> fileScanTask.length())
+        .reduce(0L, Long::sum);
+  }
+
+  private long calculateBytes(SplitsChange<IcebergSourceSplit> splitsChanges) {
+    return splitsChanges.splits().stream()
+        .map(split -> calculateBytes(split))
+        .reduce(0L, Long::sum);
+  }
+
+  private void checkSplitOrStartNext() throws IOException {
+    if (currentReader != null) {
+      return;
+    }
+
+    IcebergSourceSplit nextSplit = splits.poll();
+    if (nextSplit == null) {
+      throw new IOException("No split remaining");
```

Comment:
Maybe add a comment to explain why this is the right behavior. It seems odd to me that it should throw an exception when it hasn't received new splits. Maybe that is the expectation from how `fetch` is called, but it is definitely concerning to see.

Reply:
This is a great question. I looked into Flink code again. Throwing an IOException will cause the fetcher to be closed so that we don't have to poll the fetcher again. that is not a problem, because new split later will create new fetcher.

This is the `FetchTask` that calls `splitReader.fetch`. We can see that the return value can't be null. I agree that this is a little weird behavior. cc @tweise

```
    public boolean run() throws IOException {
        try {
            if (!isWakenUp() && lastRecords == null) {
                lastRecords = splitReader.fetch();
            }

            if (!isWakenUp()) {
                // The order matters here. We must first put the last records into the queue.
                // This ensures the handling of the fetched records is atomic to wakeup.
                if (elementsQueue.put(fetcherIndex, lastRecords)) {
                    if (!lastRecords.finishedSplits().isEmpty()) {
                        // The callback does not throw InterruptedException.
                        splitFinishedCallback.accept(lastRecords.finishedSplits());
                    }
                    lastRecords = null;
                }
            }
        }
```

Reply:
aded a method comment section to explain the IOException behavior

Reply:
This is probably better as an inline comment.

Reply:
changed to inline comment

Reply:
How about making it a blocking poll that can be waken up by `#wakeUp` per 
https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitReader.java#L37-L48

It's ok to keep it simple at first though.

Reply:
@yittg that is a good question. In theory, I agree it is a good behavior. I haven't figured a good way to implement the wakeUp behavior

Say there is no split available from the queue. if fetch() method doesn't return, it would need to poll the `splits` queue to check for new split in some loop with sleep. if the wakerUp is called and set the wakeUp flag for the fetch method to break out of the loop and return an empty `RecordsWithSplitIds` with no records and no finished splits.

We can't use wait-notify to signal btw wakeUp and fetch. Otherwise, fetch method won't do a splits queue polling and discover new splits during wait. That is also undesirable.

Maybe there is other way I haven't thought about.

Reply:
My colleague suggested a good way. We can just return an empty result in this case. it will cause fetcher to be idle and SplitFetcherManager will closes idle fetcher. At least, it is like a normal flow (than IOException). I pushed a commit for the change.

Reply:
SGTM roughly, thanks

Reply:
Forgot to mention. For the change of return an empty result for no split case, it has been validated by the unit test in the uber draft PR, which contains MiniCluster unit tests for both batch and streaming executions. https://github.com/apache/iceberg/pull/2105